usr/sbin/p11sak/p11sak.c

```diff
@@ -887,17 +908,21 @@ static CK_RV tok_key_gen(CK_SESSION_HANDLE session, CK_ULONG keylength,
 
     CK_ATTRIBUTE tmplt[] = {
             // boolean attrs
-            { CKA_TOKEN, &a_token, bs }, { CKA_PRIVATE, &a_private, bs }, {
-            CKA_MODIFIABLE, &a_modifiable, bs }, { CKA_DERIVE, &a_derive, bs },
-            { CKA_SENSITIVE, &a_sensitive, bs }, {
-            CKA_ENCRYPT, &a_encrypt, bs }, { CKA_DECRYPT, &a_decrypt, bs }, {
-                    CKA_SIGN, &a_sign, bs }, {
-            CKA_VERIFY, &a_verify, bs }, { CKA_WRAP, &a_wrap, bs }, {
-            CKA_UNWRAP, &a_unwrap, bs },
+            { CKA_TOKEN, &a_token, bs }, 
+            { CKA_PRIVATE, &a_private, bs }, 
+            { CKA_MODIFIABLE, &a_modifiable, bs }, 
+            { CKA_DERIVE, &a_derive, bs },
+            { CKA_SENSITIVE, &a_sensitive, bs }, 
+            { CKA_ENCRYPT, &a_encrypt, bs }, 
+            { CKA_DECRYPT, &a_decrypt, bs }, 
+            { CKA_SIGN, &a_sign, bs }, 
+            { CKA_VERIFY, &a_verify, bs }, 
+            { CKA_WRAP, &a_wrap, bs }, 
+            { CKA_UNWRAP, &a_unwrap, bs },
             { CKA_EXTRACTABLE, &a_extractable, bs },
             // non-boolean attrs
-            { CKA_VALUE_LEN, &a_value_len, sizeof(CK_ULONG) }, { CKA_LABEL,
-                    label, strlen(label) } };
+            { CKA_VALUE_LEN, &a_value_len, sizeof(CK_ULONG) }, 
+            { CKA_LABEL, label, strlen(label) } };
```

Comment:
This code still sets up a template with all Boolean attributes being FALSE by default. With your changed `set_bool_attr_from_string()` the user can now explicitly set any of those attributes to TRUE or FALSE, but any attributes that the user does not specify still default to FALSE, because they are all added to the template unconditional. So this does not yet fix Issue https://github.com/opencryptoki/opencryptoki/issues/425.

You will need to use a logic like in `set_battr()`, which is used for generating asymmetric keys, that only add those Boolean attributes to the template that are specified by the user, but do not add any attributes that the user did not specify. That way, the non-specified attributes will get the default value by OCK (as we want it to be).

Reply:
Ahhh, I see you are changing this in another commit. So all is good. please ignore this comment :-)